core/cap-0046-08.md

```diff
@@ -28,70 +28,47 @@ This CAP is aligned with the following Stellar Network Goals:
-a user to cryptographically verify anything in `TransactionMetaV3`.
-`TransactionMetaV3` will store separate SHA256 hashes for ledger entry changes and
-contract events, and the hash of those hashes will be stored in `TransactionResultPairV2`.
-
-How can this be used? The `LedgerHeader` contains the hash of the
-`TransactionResultSetV2`, so the user will first gather all results from the desired
-ledger, hash them, and verify that the hash matches what's in the `LedgerHeader`.
-With `TransactionResultSetV2` verified, the user can then find the `TransactionMetaV3`
-for the transaction they're interested in, hash just contract events, combine
-that hash with the other hashes in `TransactionMetaV3` and then verify it against
-the hash stored in `TransactionResultPairV2`.
+We introduced a `TransactionMetaV3`, which has an optional
+`SorobanTransactionMeta` struct that should be set for Soroban transactions.
+This contains a `events` vector of `ContractEvent`, which will contain the
+events emitted by the host and contract.
 
-#### Host function
-The `contract_event` host function specified in [CAP-0046-03](./CAP-0046-03.md#context-host-functions)
-can be used to create events in contracts.
+### Cryptographic verification
 
-#### Stop storing TransactionResults in archives
-This CAP also removes the `TransactionResult` from the archives by using
-`TransactionHistoryResultEntryV2`. Instead, the `TransactionResult` will be
-stored in `TransactionMetaV3`, while the hash of `TransactionResult` will be
-stored in `TransactionMetaV3::hashes`. With this change, the archives will take less
-space, but it does mean that the results will need to be be pulled from the
-meta. Cryptographic verification would work just like contract events described
-above, except you would match against the `TransactionResult` hash stored in
-`TransactionMetaV3::hashes`.
+`TransactionMeta` is not hashed into the ledger, which is an issue because the
+events can't be cryptographically verified, and tests that replay historical
+transactions will not catch instances where the events emitted have changed. To
+fix this issue, `InvokeHostFunctionOp` returns the SHA-256 hash of
+`InvokeHostFunctionSuccessPreImage`, which contains the SCVal return value of
+the top level contract call, and the vector of events emitted. This makes the
+events part of the protocol. 
 
-`LedgerHeader::txSetResultHash` will also be updated to store the hash of
-`TransactionResultSetV2`.
 
-TODO: Does this work with InnerTransactionResultPair?
+### Diagnostic events
+The host can also emit diagnostic events, which are not part of the protocol and
+do not count towards metering and fees. These events are in a separate
+`diagnosticEvents` vector in `SorobanTransactionMeta` with `ContractEventType`
+`DIAGNOSTIC`. To be able to order diagnostic and non-diagnostic events,
+`diagnosticEvents` will also contain all of the protocol events. 
```

Comment:
'will also contain all of the protocol events in order of occurrence`?

Reply:
done